StripeConnect/StripeConnect/Source/Internal/Webview/ConnectWebView.swift

```diff
@@ -0,0 +1,187 @@
+import WebKit
+
+/**
+ Custom implementation of a web view that handles:
+ - Camera access
+ - Popup windows
+ - Opening email links
+ - Downloads  TODO MXMOBILE-2485
+ */
+class ConnectWebView: WKWebView {
+
+    /// Closure to present a popup web view controller.
+    /// This is required for any components that can open a popup, otherwise an assertionFailure will occur.
+    var presentPopup: ((UIViewController) -> Void)?
+
+    /// Closure that executes when the view finishes loading.
+    /// - Note: If any JS needs to be evaluated immediately after instantiation, do that here.
+    var didFinishLoading: ((ConnectWebView) -> Void)?
+
+    /// Closure that executes when `window.close()` is called in JS
+    var didClose: ((ConnectWebView) -> Void)?
+
+    /// The instance that will handle opening external urls
+    let urlOpener: URLOpener
+    
+    /// The current version for the SDK
+    let bundleVersion: String?
+    
+    init(frame: CGRect,
+         configuration: WKWebViewConfiguration,
+         // Only override for tests
+         urlOpener: URLOpener = UIApplication.shared,
+         bundleVersion: String? = StripeConnectBundleLocator.bundleVersion) {
+        self.urlOpener = urlOpener
+        self.bundleVersion = bundleVersion
+        configuration.applicationNameForUserAgent = "- stripe-ios/\(bundleVersion ?? "")"
+        super.init(frame: frame, configuration: configuration)
+        
+        // Allow the web view to be inspected for debug builds on 16.4+
+        #if DEBUG
+        if #available(iOS 16.4, *) {
+            isInspectable = true
+        }
+        #endif
+        
+        uiDelegate = self
+        navigationDelegate = self
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+// MARK: - Private
+
+private extension ConnectWebView {
+    // Opens the given navigation in a PopupWebViewController
+    func openInPopup(configuration: WKWebViewConfiguration,
+                     navigationAction: WKNavigationAction) -> WKWebView? {
+        guard let presentPopup else {
+            assertionFailure("Cannot present popup")
+            return nil
+        }
+
+        let popupVC = PopupWebViewController(configuration: configuration, navigationAction: navigationAction, urlOpener: urlOpener,
+                                             bundleVersion: bundleVersion
+        )
+        let navController = UINavigationController(rootViewController: popupVC)
+        popupVC.navigationItem.rightBarButtonItem = .init(systemItem: .done, primaryAction: .init(handler: { [weak popupVC] _ in
+            popupVC?.dismiss(animated: true)
+        }))
+
+        presentPopup(navController)
+        return popupVC.webView
+    }
+
+    // Opens the given URL in an SFSafariViewController
+    func openInAppSafari(url: URL) {
+        guard let presentPopup else {
+            assertionFailure("Cannot present popup")
+            return
+        }
```

Comment:
Optional:
Given that there are multiple callsites where we check for `presentPopup` and assertionFail, we could DRY this up into a helper:

```
private var optionalPresentPopup: ((UIViewController) -> Void)?

var presentPopup: (UIViewController) -> Void {
    get {
        assert(optionalPresentPopup != nil,  "Cannot present popup")
        return optionalPresentPopup ?? { _ in }
    }
    set {
        optionalPresentPopup = newValue
    }
}
```